### ISSUE TITLE
Foreign Key Filtering/Retrieval

### ISSUE THREAD

**Issue by cliedeman** [facebookincubator/ent#92]
Given the following schema

```
// todo.go
type Todo struct {
	ent.Schema
}

func (Todo) Fields() []ent.Field {
	return []ent.Field{
		field.Text("text"),
		field.Bool("done").Default(false),
	}
}

func (Todo) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("user", User.Type).
			Ref("todos").
			Unique(),
	}
}
```

```
// user.go
type User struct {
	ent.Schema
}

func (User) Fields() []ent.Field {
	return []ent.Field{
		field.String("name").
			Default("unknown"),
	}
}

func (User) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("todos", Todo.Type),
	}
}
```

If I run the following code
`todos, err := client.Todo.Query().All(ctx)`

There is no easy way to retrieve the user_id field without sending an additional query for each row right now.

Use Case:
Graphql Query that returns users then uses a data loader to fetch users. Ideal SQL:

```
select name, done, user_id from todo;
select name from users where user_id in (1,2,3,4);
```

Option 1:
Remove edge relation and manage fk manually.

Easy, works right now but loses a lot of the advantages especially around filtering

Option 2:
Duplicate fk column:

```
func (Todo) Fields() []ent.Field {
	return []ent.Field{
		field.Text("text"),
		field.Int("user_id_fk").StorageKey("user_id"),
		field.Bool("done").Default(false),
	}
```

(user_id column name has to be used to prevent function collisions)

```ent/todo_create.go:49:23: (*TodoCreate).SetUserID redeclared in this block
        previous declaration at ent/todo_create.go:29:6
ent/todo_update.go:72:23: (*TodoUpdate).SetUserID redeclared in this block
        previous declaration at ent/todo_update.go:41:6
ent/todo_update.go:254:27: (*TodoUpdateOne).SetUserID redeclared in this block
        previous declaration at ent/todo_update.go:223:6
```

This also breaks because we get conflicting column definitions

Option 3:
Embed a readonly fk column that can be used for queries and retrieval without joining on the users table.

```
type Todo struct {
	config `json:"-"`
	// ID of the ent.
	ID int `json:"id,omitempty"`
	// Text holds the value of the "text" field.
	Text string `json:"text,omitempty"`
	// UserIDFk holds the value of the "user_id_fk" field.
	UserIDFk int `json:"user_id_fk,omitempty"`
	// Done holds the value of the "done" field.
	Done bool `json:"done,omitempty"`
}
```

I haven't fully fleshed out this idea but a good precursor might be readonly fields. For example audit fields.

**Comment by a8m** [comment]
Filtering on edges can be done using the [edge predicates](https://entgo.io/docs/predicates/#edge-predicates).

>  Embed a readonly fk column that can be used for queries

Some edges, like M2M have a join table. Hence, FK is useless in these cases. For example:
If you have a `users<-->groups` relation, you want to be able to get the groups of users easily and not the FK in the join table.  

About edge retrieving - this is part of the [eager-loading](#91) feature and I think it's better to move the discussion to one place. Are you fine with closing this issue and continue the discussion in #91?

**Comment by cliedeman** [comment]
Hi, thanks for the response.

Edge predicates is an option but isn't the extra join a waste in this case?

```
todos, err := client.Todo.Query().Where(
		todo.HasUserWith(user.IDIn(1, 2))).All(ctx)
```

emits
```
SELECT DISTINCT `todos`.`id`, `todos`.`text`, `todos`.`done` 
    FROM `todos` WHERE `todos`.`user_id` IN (
        SELECT `id` FROM `users` WHERE `users`.`id` IN (?, ?)
    )
```


Eager loading would solve some of these issues but also introduces an extra join.

I am looking for a solution where I don't need that join. Also I am happy if you simplicity sake the library does not support this or chooses not too but it seems like a very common use case to me

**Comment by a8m** [comment]
Thanks for your comment @cliedeman and sorry for the delay (holiday season until Oct 23rd).

> Edge predicates is an option but isn't the extra join a waste in this case?

The FK doesn't always exist on the same table (M2M, O2M and some O2O) and JOIN (or sub-query) is required. It's also required in all cases when querying other fields (besides IDs). 

However, I agree that cases like this should be optimized in `entc`. But in general, FKs are storage internal details and we're not planning to expose them as part of the API. 

If a user wants to apply an SQL specific predicate he can do it today. Taking you example:
```go
pets := client.Pet.
	Query().
	Where(predicate.Pet(func(s *sql.Selector) {
		// Assuming you know that "owner_id" exists
		// on the "pets" table.
		s.Where(sql.InInts(pet.OwnerColumn, 1, 2, 3))
	})).
	AllX(ctx)
// Emits: 
// SELECT DISTINCT `pets`.`id`, `pets`.`name` FROM `pets` WHERE `owner_id` IN (?, ?, ?)
// args=[1 2 3]
```

> Eager loading would solve some of these issues but also introduces an extra join.

Like above, in some cases JOIN is required. Avoiding JOIN when the edge exists on the same table is an implementation optimization, but I don't think we should change the public API for this.

**Comment by cliedeman** [comment]
> The FK doesn't always exist on the same table (M2M, O2M and some O2O) and JOIN (or sub-query) is required. It's also required in all cases when querying other fields (besides IDs).

This is interesting to me because I 
Yes my use case is verify specific. Normally you would just handle this with eager loading and call it a day. But with graphql resolvers it's a bit awkward knowing you will do that query anyway.

Some example code:

```
func (r *queryResolver) Todos(ctx context.Context) ([]*Todo, error) {
	client, _ := EntClientFromContext(ctx)

	dbTodos, _ := client.Todo.Query().Where(
		todo.HasUserWith(user.IDIn(1, 2))).All(ctx)

	var todos []*Todo

	for _, t := range dbTodos {

		todos = append(todos, &Todo{
			ID: t.ID,
			Text: t.Text,
			Done: t.Done,
			// Pass user id to child resolver
			User: &User{
				ID: t.UserID,// This field does not exist
			},
		})
	}

	return todos, nil
}

func (r *todoResolver) User(ctx context.Context, todo *Todo) (*User, error) {
	loader := ctx.Value(userLoaderKey)
	return loader.(*UserLoader).Load(todo.User.ID)
}
```

(Example using [gqlgen](https://gqlgen.com/) with ent)

So here we query todos, and then use a data loader to batch fetch the user objects.
So in this case eager loading would fix the issue but would do an extra join. A micro optimisation for sure but it's good to know what the library will and will not support. 

> However, I agree that cases like this should be optimized in entc. But in general, FKs are storage internal details and we're not planning to expose them as part of the API.

This is a bit of a contradiction :). So I'm not sure what you mean.

> Like above, in some cases JOIN is required. Avoiding JOIN when the edge exists on the same table is an implementation optimization, but I don't think we should change the public API for this.

Happy with the filtering example you are presenting here. If the code needs to be optimized so much for this to matter it should probably be hand written sql anyway. To be honest I don't have a good solution for the retrieval side. I think adding a foreign key field is bad idea and will be confusing 

E.g.

```
type Todo struct {
  User User,
  UserId: string
}
```

So in summary. I would have liked to have foreign keys retrieved when the Todo table is queried but I can't see a clean way to do it that won't mess with the api. Feel free to close

**Comment by a8m** [comment]
> So in this case eager loading would fix the issue but would do an extra join. 

What I tried to say is: with minor change in `entc` (ent codegen) the code can be optimized, and subquery can be avoided for cases like:
```go
dbTodos, _ := client.Todo.Query().Where(todo.HasUserWith(user.IDIn(1, 2))).All(ctx)
```
(Because `user_id` exists in the `todos` table. This is a M2O relation. **It will be addressed soon**).

> This is a bit of a contradiction :). So I'm not sure what you mean.
> Happy with the filtering example you are presenting here.

Again, M2O relations (and some O2O cases) will be improved by `entc`.  Also, the custom filtering won't be needed after this change in order to achieve better performance. 

> So in summary. I would have liked to have foreign keys retrieved when the Todo table is queried but I can't see a clean way to do it that won't mess with the api.

Give a look on the [eager loading](#91) proposal. This will make edge loading much simpler and faster.

BTW, it's not part of our roadmap, but we want to auto-bind this into `gqlgen` as well. I'll try to keep this issue updated with our solution.

**Comment by jvitoroc** [comment]
Any updates?

**Comment by Zn4rK** [comment]
@a8m:
> However, I agree that cases like this should be optimized in `entc`. But in general, FKs are storage internal details and we're not planning to expose them as part of the API.

Is it possible to re-open this issue?

I have a use case where I have a tree that's a O2M. Almost exactly like the one in the documentation. https://entgo.io/docs/schema-edges/#o2m-same-type.

With the key difference that I've manually specified the `StorageKey`. 

The only difference is that all nodes belongs to another thing. So I have an additional `edge.From` entry.

What I essentially want to do is to fetch all items that belongs to that thing, and have the client figure out the tree. For this specific use case it makes more sense to let the client handle it. But to do that I need access to the primary key.

I also have another use case, where I am utilising the relationship between the nodes.

Maybe FKs could be optionally exposed if they have a name specified?

**Comment by a8m** [comment]
Hey @Zn4rK!

> What I essentially want to do is to fetch all items that belongs to that thing

Are you trying to avoid the sub-query? If that's the case, I agree we need to optimize this.

> Maybe FKs could be optionally exposed if they have a name specified?

I don't really want to expose the FKs in the default codegen. If the reason is perf-tuning, we need to address these optimizations in the default behavior. 

However, I think we can allow anyone who wants it, to get it using an extension for the codegen - let's say using [external-templates](https://entgo.io/docs/templates/) and [edge-annotations](https://entgo.io/docs/schema-edges/#annotations).

**Comment by Zn4rK** [comment]
> Are you trying to avoid the sub-query? If that's the case, I agree we need to optimize this.

Not necessarily, but sure, it's an added bonus. The main reason is that GraphQL - which I'm using as the transport layer -  doesn't really support recursive types in queries. So I'm trying to keep it flat and let the client/webapp figure out the relations.

I essentially don't want to "QueryParent()" to figure out the parent id - since I, in theory should already know It, since I've already fetched all `Nodes` that belongs to specific `Thing`

- A `node` belongs to a `thing`
- A `node` can also belong to another `node`

So in my use case, the relationship here is not an internal implementation detail. 

> I don't really want to expose the FKs in the default codegen. If the reason is perf-tuning, we need to address these optimizations in the default behavior.

The reason is not perf-tuning in my case, even though I agree with you on this point. 

I've managed to hack it for now with an external template, as you suggested. But I'm fairly certain that there are other use cases where it would be befinitial to have access to the FK in the default codegen. But, I understand your point of view.

**Comment by llamadeus** [comment]
I'm facing a similar problem having no access to the underlying foreign keys. I'm working with two tables. One `products` table with 30k records and one `product_attributes` table with around 1.3 million records.

In order to implement a fast lookup by id I need to prefetch all database records and map them by their ids. I first fetch all products, then all product attributes and then iterate over both collections and map them by their ids.

```go
type cacheEntry struct {
	Product        *ent.Product
	AttributesById map[int]*ent.Attribute
}

cache := make(map[int]cacheEntry)
products, _ := client.Product.Query().All(ctx)
productAttributes, _ := client.Attribute.Query().All(ctx)

for _, product := range products {
	cache[product.ID] = productCacheEntry{
		Product: product,
		AttributesById: make(map[int]*ent.Attribute),
	}
}

for _, attribute := range productAttributes {
	if _, ok := cache[1]; ok { // 1 because attribute.ProductID does not exist on attribute
		cache[1].AttributesById[attribute.ID] = attribute
	}
}
```

The whole code takes about 150ms to execute.

If I instead use eager loading to fetch the products along with their attributes the code takes 10x longer, about 1.5 seconds.

```go
productsCache := make(map[int]productCacheEntry)
products, _ := client.Product.Query().WithAttributes().All(ctx)

for _, product := range products {
	productsCache[product.ID] = productCacheEntry{
		Product:        product,
		AttributesById: make(map[int]*ent.Attribute),
	}
	
	for _, attribute := range product.Edges.Attributes {
		if _, ok := productsCache[product.ID]; ok {
			productsCache[product.ID].AttributesById[attribute.ID] = attribute
		}
	}
}
```

@a8m Do you see using custom extensions being the only way in order to get this running? Or am I missing something?

**Comment by llamadeus** [comment]
I'm facing a similar problem having no access to the underlying foreign keys. I'm working with 3 tables. One `products` table with 30k records, an `attributes` table with only 3k records, and the pivot table `product_attributes` table with around 1.3 million records. Products and attributes are M2M.

In order to implement a fast lookup I need to prefetch all `product_attributes` entries and map them by their foreign keys (`product_id` and `attribute_id`). First, I fetch all product attributes and then iterate over all 1.3M records and map them by their ids.

```go
type cacheEntry struct {
	Product        *ent.Product
	AttributesById map[int]*ent.ProductAttribute
}

cache := make(map[int]cacheEntry)
products, _ := client.Product.Query().All(ctx)
productAttributes, _ := client.ProductAttribute.Query().All(ctx)

for _, product := range products {
	cache[product.ID] = productCacheEntry{
		Product: product,
		AttributesById: make(map[int]*ent.Attribute),
	}
}

for _, attribute := range productAttributes {
	if _, ok := cache[1]; ok { // 1 because attribute.ProductID does not exist on attribute
		cache[1].AttributesById[attribute.ID] = attribute
	}
}
```

The whole code takes about 150ms to execute.

If I instead used eager loading to fetch the products along with their attributes the code takes 10x longer, about 1.5 seconds.

```go
productsCache := make(map[int]productCacheEntry)
products, _ := client.Product.Query().WithAttributes().All(ctx)

for _, product := range products {
	productsCache[product.ID] = productCacheEntry{
		Product:        product,
		AttributesById: make(map[int]*ent.Attribute),
	}
	
	for _, attribute := range product.Edges.Attributes {
		if _, ok := productsCache[product.ID]; ok {
			productsCache[product.ID].AttributesById[attribute.ID] = attribute
		}
	}
}
```

@a8m Do you see using custom extensions being the only way in order to get this running? Or am I missing something?
